Add assertions for Equality when type implements IComparable #172

rubenrorije opened the pull request:
During creating this pull request I stumbled upon the following points of attention.

1. In the original code `Be` checks for reference equality, but `NotBe` does not. I changed the `NotBe` to also check reference equality.
2. It would be useful to differ the error messages that are shown to the user. Ideally there would need to be three different error messages: for reference equality, for 'normal' equality, and one for comparison. I am not sure how to change the code in such a way that different error messages are returned for the different cases. I can also imagine that a discussion about the correct wording might be necessary first.
For example (NotBe):
    - Did not expect subject to point to the same reference ...
    - Did not expect subject to return 0 for comparison ...
    - Did not expect subject to be equal to ...
3. I added tests for reference equality, which seemed to be missing.
4. I added tests for the new behaviour, but am not sure whether it conforms to the standards and especially whether the naming is that intuitive, I could use an exta pair of eyes.

Part of #172

dennisdoomen commented on Src/FluentAssertions/Numeric/ComparableTypeAssertions.cs:
```diff
@@ -35,7 +36,7 @@ public ComparableTypeAssertions(IComparable<T> value) : base(value)
         {
             Execute.Assertion
-                .ForCondition(ReferenceEquals(Subject, expected) || (Subject.CompareTo(expected) == Equal))
+                .ForCondition(ReferenceEquals(Subject, expected) || ((Subject.CompareTo(expected) == Equal) && Subject.IsSameOrEqualTo(expected)))
```
I think we should _only_ use `ISameOrEqual` since that will ensure `Be` behaves like other methods

rubenrorije replied:
Do I understand correctly that the easiest solution then would be to just remove these methods `Be` and `NotBe` altogether from `ComparableTypeAssertions`? It would fall back on the ObjectAssertions. This would have the downside, though, of not checking the implementation of `CompareTo` anymore.

dennisdoomen replied:
Yep. That's why we [suggested](https://github.com/fluentassertions/fluentassertions/issues/172#issuecomment-59910838) adding another method for that.

rubenrorije replied:
From the thumbs up on my suggestion I (wrongly) assumed that was the way to go. Maybe I jumped the gun on this.

dennisdoomen replied:
You're almost there, aren't you?

rubenrorije replied:
Now I am confused. My understanding is that you only want Equality and no Reference Equality and Comparison checks?
What I implemented is checking all three, to make sure that people who are depending on the behaviour that the comparison is checked is not implicitly removed in a new version. Most of the implementation is checking these three cases.

dennisdoomen replied:
Reference quality is usually checked with `BeSameAs`.

dennisdoomen commented on Src/FluentAssertions/Numeric/ComparableTypeAssertions.cs:
```diff
@@ -125,7 +127,7 @@ public AndConstraint<ComparableTypeAssertions<T>> Be(T expected, string because
         {
             Execute.Assertion
-                .ForCondition(Subject.CompareTo(unexpected) != Equal)
+                .ForCondition(!ReferenceEquals(Subject, unexpected) && (Subject.CompareTo(unexpected) != Equal || !Subject.IsSameOrEqualTo(unexpected)))
```
Same here. Just use `!IsSameOrEqualTo`

rubenrorije commented:
Is this what you had in mind?

dennisdoomen commented:
Any change you'll continue with this?

rubenrorije commented:
Removed most code I added, because of the changes requested. Now the code only checks equality and does not check comparison. Which meant a lot of the tests where no longer needed. Now the new implementation is identical to the one in ObjectAssertions.

dennisdoomen commented:
Your PR now also includes all the other's commits that were added recently. You either need to re-base or create a new branch and cherry-pick/squash your own changes.

rubenrorije commented:
Started over in a new pull request to tidy up commits
